File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteMiddle = function(head) {
  // Use slow and fast pointers, also keep track of prev node of slow
  // Find middle node, then set prev.next to slow.next to delete middle node
  // Return head
  
  if (!head.next) return head.next;
  
  let prev = ListNode();
  let slow = head, fast = head;
  
  while (fast && fast.next) {
    prev = slow;
    slow = slow.next;
    fast = fast.next.next;
  }
  
  prev.next = slow.next;
  
  return head;
};

File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
      # Use slow and fast pointers
      # Find middle node, and keep track of prev of slow node
      # O(n/2) time bc we are only going through half the nodes w fast pointer
      
      if head.next is None:
        return head.next
      
      prev = ListNode()
      slow = fast = head
      
      while fast and fast.next:
        prev = slow
        slow = slow.next
        fast = fast.next.next
      
      # Delete middle node by setting prev.next to slow.next
      prev.next = slow.next
      
      return head